/* 
 * The Biomechanical ToolKit
 * Copyright (c) 2009-2010, Arnaud Barr√©
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 *     * Redistributions of source code must retain the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer.
 *     * Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *     * Neither the name(s) of the copyright holders nor the names
 *       of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written
 *       permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef AbstractView_h
#define AbstractView_h

#include "ui_AbstractView.h"

class AbstractView : public QWidget, public Ui::AbstractView
{
  Q_OBJECT
  
public:
  AbstractView(QWidget* parent = 0);
  // ~AbstractView(); // Implicit
  // AbstractView(const AbstractView&);  // Implicit
  // AbstractView& operator=(const AbstractView&); // Implicit.
  
  virtual AbstractView* clone() const;
  
public slots:
  void setFocus(Qt::FocusReason reason);
    
private slots:
  void close();
  void splitHorizontally();
  void splitVertically();
  void setCurrentIndex(int idx);
  
signals:
  void closeTriggered(AbstractView* sender);
  void splitTriggered(AbstractView* sender, int direction);
  
private:
  void finalizeUi();
};
    
#endif // AbstractView_h
